### ISSUE TITLE
tuple length type narrowing

### ISSUE THREAD

**Issue by GBeauregard** [microsoft/pyright#2474]
**Describe the bug**
pyright should support type narrowing based on tuple length to avoid the false positive in the following code snippet:

```python
from typing import Union

MultiTuple = Union[tuple[object], tuple[object, object]]

def process_tuples(tup: MultiTuple) -> object:
    if len(tup) == 1:
        return tup[0]
    else:
        first, second = tup
        return second
```

```console
  /home/gregory/Downloads/testcase/pytest/tuplesize.py:9:9 - error: Tuple size mismatch: expected 2 but received 1 (reportGeneralTypeIssues)
```

This came up in the rich codebase, and there's a corresponding mypy issue with an open PR: https://github.com/python/mypy/issues/1178

**Expected behavior**
pyright should be able to type narrow tuples based on len()

**VS Code extension or command-line**
pyright 1.1.180 Linux commandline
